programs/bpf_loader/src/lib.rs

```diff
@@ -397,95 +407,110 @@ fn process_loader_upgradeable_instruction(
     use_jit: bool,
 ) -> Result<(), InstructionError> {
     let log_collector = invoke_context.get_log_collector();
-    let program_id = invoke_context.transaction_context.get_program_key()?;
-    let keyed_accounts = invoke_context.get_keyed_accounts()?;
+    let transaction_context = &invoke_context.transaction_context;
+    let instruction_context = transaction_context.get_current_instruction_context()?;
+    let program_id = transaction_context.get_program_key()?;
 
     match limited_deserialize(instruction_data)? {
         UpgradeableLoaderInstruction::InitializeBuffer => {
-            let buffer = keyed_account_at_index(keyed_accounts, first_instruction_account)?;
-
-            if UpgradeableLoaderState::Uninitialized != buffer.state()? {
+            let mut buffer = instruction_context
+                .try_borrow_account(transaction_context, first_instruction_account)?;
+            if UpgradeableLoaderState::Uninitialized != buffer.get_state()? {
                 ic_logger_msg!(log_collector, "Buffer account already initialized");
                 return Err(InstructionError::AccountAlreadyInitialized);
             }
 
-            let authority = keyed_account_at_index(keyed_accounts, first_instruction_account + 1)?;
-
-            buffer.set_state(&UpgradeableLoaderState::Buffer {
-                authority_address: Some(*authority.unsigned_key()),
-            })?;
+            let authority_address = Some(*transaction_context.get_key_of_account_at_index(
+                instruction_context.get_index_in_transaction(first_instruction_account + 1)?,
+            )?);
+            buffer.set_state(&UpgradeableLoaderState::Buffer { authority_address })?;
         }
         UpgradeableLoaderInstruction::Write { offset, bytes } => {
-            let buffer = keyed_account_at_index(keyed_accounts, first_instruction_account)?;
-            let authority = keyed_account_at_index(keyed_accounts, first_instruction_account + 1)?;
```

Comment:
This is probably an error order mismatch, we used to check the authority account was present up front, but now check it much later after other checks